### ISSUE TITLE
[DAG] SDPatternMatch - match integer min/max patterns as well the existing ISD node

### ISSUE THREAD

**Issue by RKSimon** [llvm/llvm-project#108218]
We already have matchers for ISD::SMAX/SMIN/UMAX/UMIN nodes, but it'd be useful if we could match icmp+select patterns as well, preferably with the same existing m_SMax/m_SMin/m_UMax/m_UMin helpers.

This would be similar to the the IR PatternMatch MaxMin_match helpers which match the min/max intrinsics and the cmp+select patterns.

**Comment by AmrDeveloper** [comment]
Interested to work on it

**Comment by nikic** [comment]
What's the motivation for this? The IR PatternMatch behavior is mostly a leftover from the time before we introduced min/max intrinsics. I wouldn't want SDAG pattern match to repeat that without good cause.

**Comment by RKSimon** [comment]
We have several patterns (add/sub sat, abd etc.) where we want to match min/max patterns but we might not have created a min/max node due to legality etc. (either in SelectionDAGBuilder::visitSelect or DAGCombiner)

**Comment by nikic** [comment]
I see. Do you think that *all* places using those patterns would also want to match the select form?

**Comment by RKSimon** [comment]
The uses so far in DAGCombine - yes - but to avoid problems in the future we could just create m_SMaxLike (etc) sibling matchers that handle all cases (similar to how we have m_SExt and m_SExtLike - m_SExtLike matches both sext + zext_nneg cases).

**Comment by AmrDeveloper** [comment]
> Hi @AmrDeveloper, are you still working on this?

Yes i was working on other PR's, will continue on this this week end

**Comment by AmrDeveloper** [comment]
I plan to create a diff after fixing CI on this pr https://github.com/llvm/llvm-project/pull/111573, but if you are ready to work on it now, you can take it and i will go for different one 

@c8ef
